ui/dashboards/src/components/SaveDashboardButton/SaveDashboardButton.tsx

```diff
@@ -0,0 +1,88 @@
+// Copyright 2023 The Perses Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import { useState } from 'react';
+import { Button, ButtonProps } from '@mui/material';
+import { DashboardResource, isRelativeTimeRange } from '@perses-dev/core';
+import { useTimeRange } from '@perses-dev/plugin-system';
+import {
+  useDashboard,
+  useEditMode,
+  useSaveChangesConfirmationDialog,
+  useTemplateVariableActions,
+  useTemplateVariableStore,
+} from '../../context';
+
+export interface SaveDashboardButtonProps extends Pick<ButtonProps, 'fullWidth'> {
+  onSave?: (entity: DashboardResource) => Promise<DashboardResource>;
+  isReadonly: boolean;
+  variant?: 'contained' | 'text' | 'outlined';
+}
+
+export const SaveDashboardButton = ({ onSave, isReadonly, variant = 'contained' }: SaveDashboardButtonProps) => {
+  const [isSavingDashboard, setSavingDashboard] = useState<boolean>(false);
+  const { dashboard } = useDashboard();
+  const { variableDefaultValuesUpdated } = useTemplateVariableStore();
+  const { setVariableDefaultValues } = useTemplateVariableActions();
+  const { timeRange } = useTimeRange();
+  const { setEditMode } = useEditMode();
+  const { openSaveChangesConfirmationDialog, closeSaveChangesConfirmationDialog } = useSaveChangesConfirmationDialog();
+
+  const onSaveButtonClick = () => {
+    setVariableDefaultValues();
+    const timeRangeUpdated = isRelativeTimeRange(timeRange) && dashboard.spec.duration !== timeRange.pastDuration;
+
+    // Save dashboard if active timeRange from plugin-system is relative and different than currently saved
+    if (timeRangeUpdated || variableDefaultValuesUpdated) {
+      openSaveChangesConfirmationDialog({
+        onSaveChanges: (variableDefinitions, saveDefaultTimeRange, saveDefaultVariables) => {
+          if (isRelativeTimeRange(timeRange) && saveDefaultTimeRange === true) {
+            dashboard.spec.duration = timeRange.pastDuration;
+          }
+          if (saveDefaultVariables === true) {
+            dashboard.spec.variables = variableDefinitions;
```

Comment:
so i think we should be calling `setVariableDefaultValues` here because otherwise the json and store will be out of sync

```suggestion
            const variables = setVariableDefaultValues();
            dashboard.spec.variables = variables;
```

Reply:
To return the new variable definitions, you can do this in the store: 

```suggestion
        setVariableDefaultValues: () => {
            const variableDefinitions = get().variableDefinitions;
            const variableState = get().variableDefinitions;
            const updatedVariables = produce( ... ); 
            set((state) => { state.variableDefinitions = updatedVariables })
            return updatedVariables;
        }
```

That way you don't need to call `useTemplateVariableDefinitions` in the save changes dialog too

Reply:
Makes a lot of sense and much better than the hacky way I did it before! This is updated now, I'm still adjusting the checkSavedVariablesStatus util, and clarifying with Mohit about behavior after adding a brand new ListVariable (see failing e2e), but otherwise this is ready for another review